executor/tiflashtest/tiflash_test.go

```diff
@@ -1453,6 +1453,78 @@ func TestMPPMemoryTracker(t *testing.T) {
+
+	// tiflash_compute_dispatch_policy is global variable.
+	tk.MustExec("set @@session.tiflash_compute_dispatch_policy = 'consistent_hash';")
+	tk.MustQuery("select @@tiflash_compute_dispatch_policy").Check(testkit.Rows("consistent_hash"))
+	tk.MustExec("set @@session.tiflash_compute_dispatch_policy = 'round_robin';")
+	tk.MustQuery("select @@tiflash_compute_dispatch_policy").Check(testkit.Rows("round_robin"))
+	err = tk.ExecToErr("set @@session.tiflash_compute_dispatch_policy = 'error_dispatch_policy';")
+	require.Error(t, err)
+	require.Equal(t, "unexpected tiflash_compute dispatch policy, expect [consistent_hash round_robin], got error_dispatch_policy", err.Error())
+
+	// Invalid values.
+	err = tk.ExecToErr("set global tiflash_compute_dispatch_policy = 'error_dispatch_policy';")
+	require.Error(t, err)
+	require.Equal(t, "unexpected tiflash_compute dispatch policy, expect [consistent_hash round_robin], got error_dispatch_policy", err.Error())
+	err = tk.ExecToErr("set global tiflash_compute_dispatch_policy = '';")
+	require.Error(t, err)
+	require.Equal(t, "unexpected tiflash_compute dispatch policy, expect [consistent_hash round_robin], got ", err.Error())
+	err = tk.ExecToErr("set global tiflash_compute_dispatch_policy = 100;")
+	require.Error(t, err)
+	require.Equal(t, "unexpected tiflash_compute dispatch policy, expect [consistent_hash round_robin], got 100", err.Error())
+
+	tk.MustExec("create table t(c1 int)")
+	tk.MustExec("alter table t set tiflash replica 1")
+	tb := external.GetTableByName(t, tk, "test", "t")
+	err = domain.GetDomain(tk.Session()).DDL().UpdateTableReplicaInfo(tk.Session(), tb.Meta().ID, true)
+	require.NoError(t, err)
+	tk.MustExec("set @@session.tidb_isolation_read_engines=\"tiflash\"")
+
+	err = tiflashcompute.InitGlobalTopoFetcher(tiflashcompute.TestASStr, "", "", false)
+	require.NoError(t, err)
+
+	useASs := []bool{true, false}
+	// Valid values.
+	for _, useAS := range useASs {
+		config.UpdateGlobal(func(conf *config.Config) {
+			conf.UseAutoScaler = useAS
+		})
+		validPolicies := tiflashcompute.GetValidDispatchPolicy()
+		for _, p := range validPolicies {
+			tk.MustExec(fmt.Sprintf("set global tiflash_compute_dispatch_policy = '%s';", p))
+			tk1 := testkit.NewTestKit(t, store)
+			tk1.MustExec("use test")
+			tk1.MustQuery("select @@tiflash_compute_dispatch_policy").Check(testkit.Rows(p))
+			failpoint.Enable("github.com/pingcap/tidb/store/copr/testWhichDispatchPolicy", fmt.Sprintf(`return("%s")`, p))
+			err = tk1.ExecToErr("select * from t;")
+			if useAS {
+				// Expect error, because TestAutoScaler return empty topo.
+				require.Contains(t, err.Error(), "Cannot find proper topo from AutoScaler")
+			} else {
+				// This error message means we use PD instead of AutoScaler.
+				require.Contains(t, err.Error(), "tiflash_compute node is unavailable")
+			}
+		}
+	}
+	failpoint.Disable("github.com/pingcap/tidb/store/copr/testWhichDispatchPolicy")
```

Comment:
use defer func to move it before for loop, once require failed, it failpoint path can still be cleaned